### PR TITLE
logstorage: Adds option to write logs in gzip format

### PR DESCRIPTION
**Pull request by LiquidityC**
Adds functionality to allow storing offline logs in gzipped files instead of standard dlt files.

This change got more invasive then I would have liked. But it can't be helped. Since I'm quite new to the project I wouldn't mind some extra scrutiny.

In particular in regards to the `dlt_logstorage_get_idx_of_log_file()` function which got a major overhaul. Perhaps I'm under-thinking this logic?

Currently the setting `GzipCompression=1` is handled as an int. Is this something where we'd like to introduce a boolean settings type? It could still be 0/1 but we can parse it to a boolean instead perhaps? I don't mind putting in some extra work if it helps get this through. We (my company) would really benefit from it.

### CODE REVIEW

File: src/offlinelogstorage/dlt_offline_logstorage.c
Comment (alexmohr) [inline src/offlinelogstorage/dlt_offline_logstorage.c:1211]: You should add all new options into the example configuration, so they are documented

File: CMakeLists.txt
Comment (alexmohr) [inline CMakeLists.txt:107]: can we make this optional? i.e. only require zlib if a cmake option is set to on. Some systems might not have zlib and do not need it either so making this optional would make sure the build is still working there

File: src/offlinelogstorage/dlt_offline_logstorage_behavior.c
Comment (alexmohr) [inline src/offlinelogstorage/dlt_offline_logstorage_behavior.c:66]: maybe strncat should be used here and in line 114 to make sure we don't overflow

File: src/offlinelogstorage/dlt_offline_logstorage.c
Comment (LiquidityC) [reply src/offlinelogstorage/dlt_offline_logstorage.c:1211]: Added them to `doc/dlt_offline_logstorage.md`

File: src/offlinelogstorage/dlt_offline_logstorage_behavior.c
Comment (LiquidityC) [reply src/offlinelogstorage/dlt_offline_logstorage_behavior.c:66]: Agree. This required some refactoring of the entire function otherwise it wouldn't have served any point.

**Comment by LiquidityC** [comment]
Fixed your comments. The *optional compile* possibility introduces a lot of `#ifdef` stuff. I'm not certain what the preferred approach is on these. I'm not a fan but also couldn't see any obvious alternatives.

**Comment by alexmohr** [comment]
> Fixed your comments. The _optional compile_ possibility introduces a lot of `#ifdef` stuff. I'm not certain what the preferred approach is on these. I'm not a fan but also couldn't see any obvious alternatives.

That's something for @michael-methner to decide as I actually do not have any permissions here. Just stumbled upon your PR and was interested in the changes

File: src/offlinelogstorage/dlt_offline_logstorage.c
Comment (minminlittleshrimp) [inline src/offlinelogstorage/dlt_offline_logstorage.c:1070]: I do not refer a warning here, I think LOG_NOTICE or LOG_INFO is enough.

Comment (LiquidityC) [reply src/offlinelogstorage/dlt_offline_logstorage.c:1070]: All other instances of config flags omitted and similar use warnings hence I chose that. But in this particular case I really think it deserves a warning.

You don't agree that a user trying to enable something through config that isn't available merits a warning? The user is actively attempting to enable a feature that is not available.

File: src/offlinelogstorage/dlt_offline_logstorage_behavior.c
Comment (minminlittleshrimp) [inline src/offlinelogstorage/dlt_offline_logstorage_behavior.c:1061]: Missing close bracket leads to error compiling.
Maybe line 1061 should be included under ifdef?!

File: src/offlinelogstorage/dlt_offline_logstorage_behavior.c
Comment (minminlittleshrimp) [inline src/offlinelogstorage/dlt_offline_logstorage_behavior.c:728]: As far as I know, gzfwrite is added since 2017.
Hence, using it here may cause incompatible to some old machines (Ex: Ubuntu 16). Could use please consider using gzwrite as an alternative?

File: src/offlinelogstorage/dlt_offline_logstorage.c
Comment (minminlittleshrimp) [reply src/offlinelogstorage/dlt_offline_logstorage.c:1070]: OK, I agree. It sounds reasonable to me now :)

File: src/offlinelogstorage/dlt_offline_logstorage_behavior.c
Comment (LiquidityC) [reply src/offlinelogstorage/dlt_offline_logstorage_behavior.c:728]: That is an alternative. But is this really an issue? What's the use case for building dlt-daemon on a 7 year old out of service and support linux distribution?

Comment (michael-methner) [reply src/offlinelogstorage/dlt_offline_logstorage_behavior.c:728]: Using this API is fine with our use cases. So we can keep it.
Ubuntu 16.04 is still supported on paid plans, but if people really want to use it they can deactivate this new feature.

**Comment by michael-methner** [comment]
Hello @LiquidityC ,
code changes looks good to me. @minminlittleshrimp  will test your changes before we merge.

btw: Have you done performance measurements? Do you have an indication how much cpu performance is affected by turning on the compression?

Thank you very much.

**Comment by LiquidityC** [comment]
> Hello @LiquidityC , code changes looks good to me. @minminlittleshrimp will test your changes before we merge.
> 
> btw: Have you done performance measurements? Do you have an indication how much cpu performance is affected by turning on the compression?
> 
> Thank you very much.

No performance testing done no. I'll have to admit that I didn't check if you had any benchmarking setup in the project.

**Comment by LiquidityC** [comment]
I did some very basic monitoring of CPU usage with and without gzip compression enabled (same build, only config change). With gzip enabled the worst case usage stuck solidly at 1.7% when hammering the daemon with 1 message every 1ms (WriteOnMsg strategy).

Without compression the cpu usage would fluctuate between 1.0-1.7%.

Wild guess is that this fluctuation comes down to I/O buffering when gzip is disabled. But when it's enabled we have a stable cpu (higher) usage consumed by compression and we draw no benefit from I/O buffering done by the OS.

This is by no means scientific. It's just me abusing the daemon and monitoring it with `top`. But I'd say that we have similar worst case performances but we take a hit on best case performance. This will be impacted by the writing strategies utilized.

In our use case I'm certain we have the cycles to spare. Our concern and the reason for me submitting this change is 'wear leveling' and a reduction in disk usage.

**Review by minminlittleshrimp — CHANGES_REQUESTED** [review]
Please add status so that user could check when build with zip, maybe like:
message(STATUS "WITH_DLT_LOGSTORAGE_CTRL_UDEV = ${WITH_DLT_LOGSTORAGE_CTRL_UDEV}")
**message(STATUS "WITH_DLT_LOGSTORAGE_GZIP = ${WITH_DLT_LOGSTORAGE_GZIP}")**
...

File: doc/dlt_offline_logstorage.md
Comment (minminlittleshrimp) [inline doc/dlt_offline_logstorage.md:77]: Set the option to 0 causing error to read filter since dlt_logstorage_read_number will consider it is invalid:

```
[ 5543.704342]~DLT~ 3146~ERROR    ~Invalid, is not a number 
[ 5543.704343]~DLT~ 3146~ERROR    ~[ 5543.704344]~DLT~ 3146~WARNING  ~FILTER1 filter configuration is invalid 
[ 5543.704346]~DLT~ 3146~ERROR    ~dlt_daemon_logstorage_setup_internal_storage: Device connect failed
[ 5543.704347]~DLT~ 3146~INFO     ~Setting up internal offline log storage failed!
```
I highly suggest, for more comprehensive, it should be ON or OFF

Comment (LiquidityC) [reply doc/dlt_offline_logstorage.md:77]: Agree. I queried about this particularly in my original PR submission. So we'll got with boolean config values as `on`/`off`. Or do we want something dynamic where `1`/`0` and `true`/`false` work in parallel with `on`/`off`?

Comment (LiquidityC) [reply doc/dlt_offline_logstorage.md]: Fixed. I chose to document the "yes/no" values but the code will allow for 0/1 and true/false.

Comment (minminlittleshrimp) [reply doc/dlt_offline_logstorage.md]: Great. Thanks for the change. I think uppercase will look better and professional!? :) (ON/OFF).
I will test and comment later.

Comment (minminlittleshrimp) [reply doc/dlt_offline_logstorage.md]: Tested. Everything is fine now.
I just have 1 more question, since I don't have much experience with gzip. Will the logs be written to the zip file directly and in turn, or it waits for a number of logs (maybe based on file size), and then zip?
Thank you

Comment (LiquidityC) [reply doc/dlt_offline_logstorage.md]: In regards to uppercase ("ON/OFF") we can change `strncmp` to `strncasecmp`. But I'm not 100% on the availability of the `<strings.h>` header on various platforms.

Comment (minminlittleshrimp) [reply doc/dlt_offline_logstorage.md]: Thank you,
For me, your current implementation is good now.
I need to check a few things related to performance (I am working on some stress tests, e.g. many apps sending many logs in short period of time delay, etc)
I will comment the result when done.

**Comment by LiquidityC** [comment]
Squashed the commits.

**Comment by minminlittleshrimp** [comment]
Hello all,

I have done some raw stress tests from my side:

Setup:

dlt_logstorage.conf
[FILTER1]
EcuID=ECU1
FileSize=100000
File=FLASH
NOFiles=100
LogLevel=DLT_LOG_VERBOSE
GzipCompression=<1/0>

dlt.conf: enable:
```
OfflineLogstorageMaxDevices = 1
OfflineLogstorageDirPath = your_path
```

Enable $dlt-daemon
Enable Test binaries in turn
Observe and record result using Gnome system monitor 
Result:
________________________________________________________________
$ dlt-test-multi-process -m 100000 -p 100 -t 1 -l 1000 -d 10 -g
Setup done. Listening. My pid: 19858
**without gzip: 0.25 ~ 0.38 ~ 0.5%CPU
with gzip: 0.37 ~ 0.5 ~ 0.63%CPU**
________________________________________________________________
$ dlt-test-stress -123
Starting stress test1...
* Register   3000 contexts...
* Unregister 3000 contexts...
Finished stress test1 

Starting stress test2... 
* Creating 64 Threads, each of them registers one context,
  sending one log message, then unregisters the context
Finished stress test2 

Starting stress test3... 
* Logging raw data, up to a size of 512
Finished stress test3 

**without gzip: max 1.37%CPU
with gzip: max 1.38%CPU**
________________________________________________________________
$ dlt-test-stress-user -n 100000 -d 1 -s 10000
**without gzip: max 2.18%CPU 10MB 100 ~ 700KiB/s
with gzip: max 2.85%CPU 10.6MB stable 290 ~ 300 KiB/s**
________________________________________________________________

Conclusion: Gzip feature is working fine with CPU consumption at an acceptable level
Regards

**Comment by minminlittleshrimp** [comment]
Hello @LiquidityC 

I would like to propose my suggestion for code sequential cohesion
There are 2 points could be considered to apply my patch:

1. Keep the build logs silent: Current implementation throws some warning msgs
```
[ 31%] Building C object src/daemon/CMakeFiles/dlt-daemon.dir/__/shared/dlt_user_shared.c.o
[ 32%] Building C object src/daemon/CMakeFiles/dlt-daemon.dir/__/offlinelogstorage/dlt_offline_logstorage.c.o
/home/lum3hc/work/github/PR442/dlt-daemon/src/offlinelogstorage/dlt_offline_logstorage.c: In function ‘dlt_logstorage_check_gzip_compression’:
/home/lum3hc/work/github/PR442/dlt-daemon/src/offlinelogstorage/dlt_offline_logstorage.c:1090:60: warning: unused parameter ‘value’ [-Wunused-parameter]
                                                      char *value)
                                                            ^~~~~
At top level:
/home/lum3hc/work/github/PR442/dlt-daemon/src/offlinelogstorage/dlt_offline_logstorage.c:439:16: warning: ‘dlt_logstorage_read_bool’ defined but not used [-Wunused-function]
 DLT_STATIC int dlt_logstorage_read_bool(unsigned int *boolean, char *value)
                ^~~~~~~~~~~~~~~~~~~~~~~~
[ 34%] Building C object src/daemon/CMakeFiles/dlt-daemon.dir/__/offlinelogstorage/dlt_offline_logstorage_behavior.c.o
[ 35%] Linking C executable dlt-daemon
[ 35%] Built target dlt-daemon

```
2. Beautify src code, make them procedural and sequential cohesion
```
DLT_STATIC int dlt_logstorage_check_nofiles(DltLogStorageFilterConfig *config,
                                            char *value)
{
    if ((config == NULL) || (value == NULL))
        return -1;

    return dlt_logstorage_read_number(&config->num_files, value);
}

DLT_STATIC int dlt_logstorage_check_gzip_compression(DltLogStorageFilterConfig *config,
                                                     char *value)
{
    if ((config == NULL) || (value == NULL))
        return -1;
    return dlt_logstorage_read_bool(&config->gzip_compression, value);
}

DLT_STATIC int dlt_logstorage_check_specificsize(DltLogStorageFilterConfig *config,
                                                 char *value)
{
    if ((config == NULL) || (value == NULL))
        return -1;

    return dlt_logstorage_read_number(&config->specific_size, value);
}
```

Please kindly have a look and leave your comment
My patch: [sequential_cohesion.patch](https://github.com/COVESA/dlt-daemon/files/10979271/sequential_cohesion.patch)

Thank you so much for your contribution!
Minh

**Comment by minminlittleshrimp** [comment]
Hello @LiquidityC 
Thank you for pointing out.
I agree with you, I just did not consider the reusability of the function.
Your analysis is very detail, and I feel fine with your fix.
I would like to compete my review and we will wait for @michael-methner then.
Great to work with you on this topic!

File: tests/gtest_dlt_daemon_offline_log.cpp
Comment (minminlittleshrimp) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: Hello,
Please kindly rebase and upload your unit test result onto this review comment.
Thank you in advance!

Comment (LiquidityC) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: Here is the log. I needed to fix some tests but there is one that keeps failing. I think the socket is blocking itself somehow. Perhaps this is a known problem test?

file:///home/lprobert/ws/dlt-daemon/build/Testing/Temporary/LastTest.log

Comment (LiquidityC) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: [LastTest.log.gz](https://github.com/COVESA/dlt-daemon/files/10992794/LastTest.log.gz)

File: tests/CMakeLists.txt
Comment (minminlittleshrimp) [inline tests/CMakeLists.txt:67]: Please fix indentation

File: tests/gtest_dlt_daemon_offline_log.cpp
Comment (minminlittleshrimp) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: Thank you, let me check that fail test case.

File: tests/CMakeLists.txt
Comment (LiquidityC) [reply tests/CMakeLists.txt]: Done

File: tests/gtest_dlt_daemon_offline_log.cpp
Comment (LiquidityC) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: It's not the test. It's my machine. VBox apparently is using port 8080. Let me get you a new log. I should be able to make them all pass

Comment (LiquidityC) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: TADA!

image

Comment (minminlittleshrimp) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: Great, thanks for your hard work :muscle:

Comment (LiquidityC) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: I need to do one more change. I'm forgetting about the compile flags. It will be in shortly.

Comment (LiquidityC) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: `DltLogStorageFilterConfig::gzlog` doesn't exist if `WITH_DLT_LOGSTORAGE_GZIP` isn't enabled.
I safed up most of the tests by ensuring data is cleared in the DltLogStorageConfig before usage since not doing this can cause some awkward bugs.
New commit submitted with this fix. Testresults are confirmed both with and without `WITH_DLT_LOGSTORAGE_GZIP`.

Sorry about all the confusion. Lot's of things on my table this week :smile:

Comment (minminlittleshrimp) [reply tests/gtest_dlt_daemon_offline_log.cpp:1]: Great, sooner or not DLT team will need to zero initialize those POD objects in those test suites.
Thank you so much!

**Comment by minminlittleshrimp** [comment]
@lvklevankhanh 
Please review and kindly approve for workflows!

File: CMakeLists.txt
Comment (LiquidityC) [reply CMakeLists.txt:110]: Unless gzwrite is being deprecated I see no value in alternating between the two. If that is the case we could just stick to gzwrite.

I think we could check the zlib version to prevent compiler errors but mixing around the code feels overkill.

File: src/offlinelogstorage/dlt_offline_logstorage_behavior.c
Comment (LiquidityC) [reply src/offlinelogstorage/dlt_offline_logstorage_behavior.c:728]: Feels like needless complexity we could just stick to gzwrite instead if that is the case. But as was mentioned earlier. gzfwrite has been available for quite some time. Users who want to start using gzip will most likely not have an issue with the zlib version.

Have a think about it and let me know if you want some changes.

File: CMakeLists.txt
Comment (michael-methner) [reply CMakeLists.txt:110]: Hello @LiquidityC , hello @lvklevankhanh ,
we already decided that we only want to support the new zlib version, so we can stick to gzfwrite().
I think it makes sense to add the minimum version number to the find_package check:
find_package(ZLIB 1.2.9 REQUIRED)

Comment (LiquidityC) [reply CMakeLists.txt:110]: Agree. I'll send a patch shortly.

Comment (lvklevankhanh) [reply CMakeLists.txt:113]: Agree with @michael-methner's comment.

**Comment by minminlittleshrimp** [comment]
Hello @LiquidityC 
For the next 2 weeks we actually plan to release a new minor update alpha version for dlt. Hence, we would like to merge your feature after the release. It would be great to have this new feature gzip merged in the new version.
FYI: https://github.com/COVESA/dlt-daemon/discussions/455

Many thanks,
DLT team

**Comment by LiquidityC** [comment]
@minminlittleshrimp Understandable. Just to double check. I'll just wait. Apart from possible review comments/changes there is nothing else you require from me. Right?

**Comment by minminlittleshrimp** [comment]
> @minminlittleshrimp Understandable. Just to double check. I'll just wait. Apart from possible review comments/changes there is nothing else you require from me. Right?

Sure, I think we’ve done the best so far. For now please kindly wait for the release.
Thanks for your understanding.

**Review by michael-methner — APPROVED** [review]

**Comment by michael-methner** [comment]
Hello @LiquidityC ,
thanks for the patch. It is ready to merge but we will delay the merge until we have released the next dlt version. We do not want to add features shortly before a release. Please wait for another two weeks.

**Comment by tjlingesh** [comment]
After including this patch, when I decompress .gz offline storage logs, I am getting 0 kb file size for the decompressed logs.
I am using gzip utility in my target, will that be enough? or do I need to explicitly use zlib library also in my machine/target?

**Comment by LiquidityC** [comment]
What size are the `gz` log files? What does your `dlt_logstorage.conf` look like?

**Comment by tjlingesh** [comment]
> What size are the `gz` log files? What does your `dlt_logstorage.conf` look like?

Compressed log files are around 16 KB.

dlt_logstorage.conf file looks like
[FILTER1]
LogLevel=DLT_LOG_INFO
File=test
FileSize=20000
NOFiles=1
EcuID=ECU
SyncBehavior=ON_SPECIFIC_SIZE
SpecificSize=4096

**Comment by LiquidityC** [comment]
I don't think there's anything in the log then. Your sync behavior is set to write when a certain buffer size has been reached. Are you sure the logs have been written when you are checking the file? Perhaps use a more aggressive sync behavior just to confirm that the log is actually being written?

I also see that you haven't specified any `LogAppName` or `ContextName` in your filter. I'm not sure how the log  storage behaves if these are omitted. You should probably specify at least one of these.
Is there any output from dlt-daemon when it's being started? If it's not happy with your `dlt_logstorage.conf` file there might be some info output about that on boot.

**Comment by tjlingesh** [comment]
> I don't think there's anything in the log then. Your sync behavior is set to write when a certain buffer size has been reached. Are you sure the logs have been written when you are checking the file? Perhaps use a more aggressive sync behavior just to confirm that the log is actually being written?
> 
> I also see that you haven't specified any `LogAppName` or `ContextName` in your filter. I'm not sure how the log storage behaves if these are omitted. You should probably specify at least one of these. Is there any output from dlt-daemon when it's being started? If it's not happy with your `dlt_logstorage.conf` file there might be some info output about that on boot.

By adding LogAppName and ContextName and removing sync behavior in dlt_logstorage.conf file, decompressed logs are working fine.

Thanks for the quick response.

**Comment by minminlittleshrimp** [comment]
Hello @tjlingesh 

For the sync behavior, you need to send a trigger signal, say:

# dlt-logstorage-ctrl -s

Could you please kindly try again with this command?

Regards
DLT team

File: src/offlinelogstorage/dlt_offline_logstorage.h
Comment (johnmcfarlane) [inline src/offlinelogstorage/dlt_offline_logstorage.h:56]: Hi, quick question regarding zlib use in dlt-deamon. If none of `WITH_DLT_LOGSTORAGE_GZIP`, `WITH_DLT_COREDUMPHANDLER`, or `WITH_DLT_FILETRANSFER` are defined, isn't `ZLIB_LIBRARY` undefined? And if so, how does CMake find this header (assuming it's not installed on the system), and how is zlib linked to dlt-daemon?

I'm trying to understand how this project works without zlib installed on the system already (if at all). Thanks!

Comment (LiquidityC) [reply src/offlinelogstorage/dlt_offline_logstorage.h:56]: Some of this might be obsolete now since this PR is old and AFAIC see there has been some re-writing of logstorage. ZLIB is linked with the daemon in `src/daemon/CMakeLists.txt`. Guarded by the variables you mentioned.
If ZLIB is excluded the variable is set to *empty string*.

**Pull request merged**
